arch/x86/entry.S

```diff
@@ -145,10 +145,9 @@ ENTRY(enter_usermode)
     PUSHF
 
     /* Save stack pointer onto per-cpu */
-    mov %_ASM_SP, %gs:usermode_private
+    mov %_ASM_DX, %gs:(usermode_private)
 
-    /* Move to user stack */
-    mov %_ASM_DX, %_ASM_SP
+    syscall_to_usermode
 
     /* SS + SP */
     push $__USER_DS
```

Comment:
Not related to this PR, but in general: Why do we use the user stack for this? Can't we put the IRET frame on the kernel stack and make `IRET` load the stack pointer?

I mean, it probably doesn't matter much, but with SMAP enabled this may fail or when we support threads and don't "trust" userland any more, userland will be able to mess with the IRET frame make it return to user code while preserving CPL 0.

Reply:
These are all good points. It was done this way merely for simplicity - I did not want to mess with the kernel stack and take care of it later. But, it should be done as you say sooner, than later anyway.